### PR TITLE
Add CustomJSTicker

### LINKED ISSUE #13130: Add CustomJSTicker
The asymmetry of having a `CustomJSTIckFormatter` while not offering a corresponding `CustomJSTicker` has low-level bothered me for years. However the utility of having one has come up in a few different support questions latetly, so we should go ahead and add one. 

It should be fairly straightforward to adapt existing "CustomJS" code to be callable by `Ticker.get_ticks` (keeping in mind @mattpap current work to afford imports inside CustomJS code). The main questions to decide:

`Ticker.get_ticks` expects to return both major and minor tick locations:
```
export type TickSpec<T> = {
  major: T[]
  minor: T[]
}
```
Do we want to expose this fully to CustomJS code? Or just expect users to return major ticks? Or have some simpler path to return only major ticks in the common case?

Additionally `get_ticks` accepts a number of parameters:
```
get_ticks(data_low: number, data_high: number, range: Range, cross_loc: number):
```
Do we expose all of these to the CustomJS code or leave off some of the more esoteric ones like `cross_loc`. In any case how do we pass the data to user code, maybe `cb_data`?

cc @bokeh/dev

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #13130
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

@mattpap we seem to have a fair amount of both discrepancy and duplication in the handling of  the various CustomJS models at this point. After this PR and the other PR are merged I'd like to make an issue to hammer out a refactoring plan that streamlines all the CustomJS models and makes them as consistent as possible.

### CODE REVIEW

File: bokehjs/test/unit/base.ts [hunk 1]
```diff
@@ -105,6 +105,7 @@ describe("default model resolver", () => {
       "CustomJSFilter",
       "CustomJSHover",
       "CustomJSTickFormatter",
+      "CustomJSTicker",
```
Comment (bryevdv) [inline bokehjs/test/unit/base.ts:108]: @mattpap FYI when the test for this fails it's really hard to debug without a proper object diff (there's just *"<50 line value> expected to equal <some other 50 line value>"*)

**Comment by bryevdv** [comment]
fixes for new mypy version are in a separate PR

**Comment by bryevdv** [comment]
One thing that occurred to me after making the example: I suspect many of the previous questions about limiting "too many" categorical labels would probably be fine with keeping the ticks, and only thinning the labels. However, these two things are currently implicitly linked. In the example I did save the original ticker on the grid, so that all the grid lines do show up. This seems like it would be a reasonable compromise for most cases IMO, but we can consider splitting off a separate optional `labels_ticker` in the future if it really seems needed.

File: src/bokeh/models/tickers.py [hunk 2]
```diff
@@ -78,6 +82,78 @@ class Ticker(Model):
...
+
+    '''
+
+    # explicit __init__ to support Init signatures
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    args = Dict(String, AnyRef, help="""
+    A mapping of names to Python objects. In particular those can be bokeh's models.
+    These objects are made available to the formatter's code snippet as the values of
```
Comment (mosc9575) [inline src/bokeh/models/tickers.py]:
```suggestion
    These objects are made available to the formatters code snippet as the values of
```

File: src/bokeh/models/tickers.py [hunk 3]
```diff
@@ -78,6 +82,78 @@ class Ticker(Model):
...
+    minor_code = String(default="", help="""
+    Callback code to run in the browser to compute minor tick locations for the
+    current viewport.
+
+    .. note::
+        Minor ticks are not used for categorical axes. This property will be
+        ignored when the range is a ``FactorRange``.
+
+    The ``cb_data`` parameter that is available to the callback code will contain
+    four specific fields:
```
Comment (mosc9575) [inline src/bokeh/models/tickers.py]:
```suggestion
    five specific fields:
```

File: examples/styling/plots/custom_js_ticker.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            (major_ticks[1]+major_ticks[2])/2,
+            (major_ticks[2]+end)/2,
+        ]
+    """,
+)
+
+yticker = CustomJSTicker(major_code="return ['a', 'c', 'e', 'g']")
+
+p = figure(y_range=list("abcdefg"))
+p.scatter([1,2,3,4,5], ["a","d","b","f","c"], size=30)
```
Comment (mosc9575) [inline examples/styling/plots/custom_js_ticker.py]:
```suggestion
p.scatter([1, 2, 3, 4, 5], ["a", "d", "b", "f", "c"], size=30)
```

**Review by mosc9575 — COMMENTED** [review]
I've found two typos and some missing whitespaces. Nothing really important

File: src/bokeh/models/tickers.py (hunk 2 above)
Comment (bryevdv) [reply src/bokeh/models/tickers.py]: The possessive seems correct to me, the core snippet belongs to ("is a property of") the object. Except the object is a ticker, not a formatter
```suggestion
    These objects are made available to the ticker's code snippet as the values of
```

**Comment by bryevdv** [comment]
@mattpap are you done commenting on this PR (and the other one)?

**Comment by bryevdv** [comment]
@mattpap I would like to merge this, do you have any comments?

**Review by mattpap — APPROVED** [review]
Looks good. However, I would mark the new model as experimental, to give ourselves time to resolve #13994, without having to go through unnecessary deprecations.

File: bokehjs/test/unit/base.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/test/unit/base.ts:108]: This has been an annoyance to me as well on multiple occasions. I'm working on incorporating an object diff tool into our testing framework to resolve this.

**Comment by bryevdv** [comment]
> I would mark the new model as experimental

It would strike me as an extremely odd incongruity to mark an API that is identical to all other existing "CustomJS" models as experimental. We are going to have to go through them all in any case, so I don't see the value in manufacturing a special case. FWIW my hope for #13994 was to consolidate implementation details, and possibly extend some functionality that only exists in `CustomJS` to other models, i.e. additive and/or hidden changes.

**Pull request merged**